library/nmstate_interface.py

```diff
@@ -52,8 +52,37 @@
 class AnsibleNMStateInterface(AnsibleNMState):
     def run(self):
         name = self.params["name"]
-        interface_state = get_interface_state(self.interfaces, name)
-        interface_state["state"] = self.params["state"]
+        full_interface_state = get_interface_state(self.interfaces, name)
+
+        interface_state = {}
+        for key in ["type", "name", "mtu", "ethernet", "description"]:
+            if key in full_interface_state:
+                interface_state[key] = full_interface_state[key]
+
+        state = self.params["state"]
+
+        # "present" does not need to be mentioned in the state sent to nmstate
+        if state != "present":
+            interface_state["state"] = state
+
+        if self.params["mtu"]:
+            interface_state["mtu"] = int(self.params["mtu"])
+
+        if self.params["duplex"]:
+            ethernet_state = interface_state.setdefault("ethernet", {})
+            if self.params["duplex"] != "auto":
+                ethernet_state["auto-negotiation"] = False
```

Comment:
Can you set the auto-negotiation post the duplex update? It will make it a bit clearer and simpler to understand/read.

Reply:
I do not follow. The ansible `duplex` parameter supports `auto`, `full`, and `half`. So this one parameter needs to be translated into two nmstate settings.

Reply:
Based on the code, you can first handle duplex value and then based on its content handle the auto-neg one.
Now their are mixed.

Reply:
Separation is better IMO, and note that ethernet_state = interface_state.setdefault("ethernet", {}) is repeating everywhere.

It is just a suggestion.

Reply:
changed it for now, let's see if I will like it in the future.